livekit-plugins/livekit-plugins-aws/livekit/plugins/aws/experimental/realtime/realtime_model.py

```diff
@@ -679,19 +837,29 @@ async def _handle_text_output_content_start_event(self, event_data: dict) -> Non
         log_event_data(event_data)
 
         role = event_data["event"]["contentStart"]["role"]
+        completion_id = event_data["event"]["contentStart"].get("completionId")
 
-        # CRITICAL: Create NEW generation for each ASSISTANT SPECULATIVE response
-        # Nova Sonic sends ASSISTANT SPECULATIVE for each new assistant turn, including after tool calls.
-        # Without this, audio frames get routed to the wrong generation and don't play.
+        # CRITICAL: Only close generation when completionId changes
+        # Nova Sonic sends multiple SPECULATIVE text blocks within the same completion
+        # (e.g., joke setup, then punchline). Closing on each SPECULATIVE would close
+        # the audio channel prematurely, causing audio cutoff.
         if role == "ASSISTANT":
             additional_fields = event_data["event"]["contentStart"].get("additionalModelFields", "")
             if "SPECULATIVE" in additional_fields:
-                # This is a new assistant response - close previous and create new
-                logger.debug("ASSISTANT SPECULATIVE text - creating new generation")
-                if self._current_generation is not None:
-                    logger.debug("Closing previous generation for new assistant response")
+                # Check if this is a different completion
+                if (
+                    self._current_generation is not None
+                    and self._current_generation.completion_id != completion_id
+                ):
+                    logger.debug(
+                        f"New completion detected (old={self._current_generation.completion_id}, new={completion_id})"
+                    )
                     self._close_current_generation()
-                self._create_response_generation()
+
+                # Create generation if needed
+                if self._current_generation is None:
```

Comment:
I faced an issue where a generation would be created multiple times, I believe it was because the generation was closed before this `if` statement so it would run anyway, making this an `elif` fixed it
```suggestion
                elif self._current_generation is None:
```